xline/src/storage/compact/revision_compactor.rs

```diff
@@ -0,0 +1,152 @@
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering::Relaxed},
+        Arc,
+    },
+    time::{Duration, Instant},
+};
+
+use clippy_utilities::OverflowArithmetic;
+use event_listener::Event;
+use tracing::{info, warn};
+
+use super::{Compactable, Compactor};
+use crate::revision_number::RevisionNumberGenerator;
+
+/// check for the need of compaction every 5 minutes
+const CHECK_INTERVAL: Duration = Duration::from_secs(5 * 60);
+
+/// Revision auto compactor
+#[derive(Debug)]
+pub(crate) struct RevisionCompactor<C: Compactable> {
+    /// `is_leader` indicates whether the current node is a leader or not.
+    is_leader: AtomicBool,
+    /// curp client
+    client: Arc<C>,
+    /// revision getter
+    revision_getter: Arc<RevisionNumberGenerator>,
+    /// shutdown trigger
+    shutdown_trigger: Arc<Event>,
+    /// revision retention
+    retention: i64,
+}
+
+impl<C: Compactable> RevisionCompactor<C> {
+    /// Creates a new revision compactor
+    pub(super) fn new_arc(
+        is_leader: bool,
+        client: Arc<C>,
+        revision_getter: Arc<RevisionNumberGenerator>,
+        shutdown_trigger: Arc<Event>,
+        retention: i64,
+    ) -> Arc<Self> {
+        Arc::new(Self {
+            is_leader: AtomicBool::new(is_leader),
+            client,
+            revision_getter,
+            shutdown_trigger,
+            retention,
+        })
+    }
+
+    /// perform auto compaction logic
+    async fn do_compact(&self, last_revision: Option<i64>) -> Option<i64> {
+        if !self.is_leader.load(Relaxed) {
+            return None;
+        }
```

Comment:
I recommend move this to the run loop, and use a event to notioy the loop when current node is not leader

Reply:
We don't need another event to do the same thing in that the `State` will invoke the `on_election_win` method to notify the auto compactor. If we move this to the run loop, we cannot cover this branch in our test.